edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/publisher/MetricsFileWriter.cs

```diff
@@ -28,6 +28,8 @@ public async Task PublishAsync(IEnumerable<Metric> metrics, CancellationToken ca
                     await file.WriteAsync(buffer, 0, buffer.Length, cancellationToken);
```

Comment:
>WriteAsync [](start = 31, length = 10)

Do we need to handle/log potential exceptions when writing to the file, in a similar way to what you did on the IoTHubMetricsUpload implementation?

Reply:
Any exceptions will propagate back to the worker, which will log them. The true/false is just for retry, not unexpected exceptions. The exceptions writing to a file will throw (probably permission denied) should not be retried.

Reply:
Although, I just realized the metrics are preserved in the case of an unexpected exception. I will change MetricsWorker so they are deleted instead, so if it continually fails to upload and can't retry it doesn't fill the disk.

Reply:
Got it. In this case, I would consider implementing a retry strategy for the MetricsPublisher, encapsulating the decision about which exceptions to retry or not. It would be a bit cleaner and also easier to extend it later (e.g. for additional exceptions). 

---
In reply to: [358507791](https://github.com/Azure/iotedge/pull/2140#discussion_r358507791) [](ancestors = 358507791)

Reply:
That's what the bool does. The publisher decides whether it should retry or not. True for success, false for retry, any exception for un-retryable failure.

Reply:
Sure. I understand. I was just thinking about encapsulating that on the detectionStrategy. Something similar to https://github.com/Azure/iotedge/blob/3eafe054159320ef61547c8fbadf812b767d7da1/edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/statemachine/EndpointExecutorFsm.cs#L535. But to be honest, take it with a grain of salt, since I am not fully familiar with the coding guideline yet.